haproxy receiver fails to collect data via unix socket when "stat" data exceeds 4k

### Component(s)

receiver/haproxy

### What happened?

## Description

When receiving stats via unix socket, the haproxy receiver only captures up to 4096 bytes: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.99.0/receiver/haproxyreceiver/scraper.go#L73

If the actual csv data is bigger than this, otel collector will return the following error and produce no metrics:
```
2024-04-23T19:34:33.325Z        error        scraperhelper/scrapercontroller.go:197        Error scraping metrics        {"kind": "receiver", "name": "haproxy", "data_type": "metrics", "error": "error reading stats: record on line 21: wrong number of fields", "scraper": "haproxy"}
```

## Steps to Reproduce

Setup an haproxy that listens on a unix socket for stats, and has enough frontends/backends/listeners to make its "stat" bigger than 4096 bytes. Configure otel collector to collect those stats via the unix socket (see below for config).

## Expected Result

Successfully capture stats

## Actual Result

Generates the following error:
```
2024-04-23T19:34:33.325Z        error        scraperhelper/scrapercontroller.go:197        Error scraping metrics        {"kind": "receiver", "name": "haproxy", "data_type": "metrics", "error": "error reading stats: record on line 21: wrong number of fields", "scraper": "haproxy"}
```

### Collector version

0.99.0

### Environment information

## Environment
OS: Ubuntu 20.04.6 LTS



### OpenTelemetry Collector configuration

```yaml
receivers:
  haproxy:
    endpoint: /run/haproxy/admin.sock
    collection_interval: 10s

exporters:
  debug:
    verbosity: detailed

service:
  pipelines:
    metrics:
      receivers: [haproxy]
      exporters: [debug]
```


### Log output

```shell
2024-04-23T19:48:13.794Z        error   scraperhelper/scrapercontroller.go:197  Error scraping metrics  {"kind": "receiver", "name": "haproxy", "data_type": "metrics", "error": "error reading stats: record on line 21: wrong number of fields", "scraper": "haproxy"}
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport
        go.opentelemetry.io/collector/receiver@v0.99.0/scraperhelper/scrapercontroller.go:197
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1
        go.opentelemetry.io/collector/receiver@v0.99.0/scraperhelper/scrapercontroller.go:169
2024-04-23T19:48:13.794Z        info    MetricsExporter {"kind": "exporter", "data_type": "metrics", "name": "debug", "resource metrics": 0, "metrics": 0, "data points": 0}
```


### Additional context

One of our haproxies has 43,279 bytes (178 lines) of "stats" data:
```
echo 'show stat' | socat /var/lib/haproxy/stats stdio | wc
    178     444   43279
```

And the 4096th byte is in the middle of line 21:
```
echo 'show stat' | socat /var/lib/haproxy/stats stdio | head -c 4096 | wc -l
20
```

--- a/.chloggen/haproxy_readall.yaml
+++ b/.chloggen/haproxy_readall.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: haproxyreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix reading stats larger than 4096 bytes
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [32652]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/receiver/haproxyreceiver/scraper.go
+++ b/receiver/haproxyreceiver/scraper.go
@@ -70,12 +70,10 @@ func (s *scraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
 		if err != nil {
 			return pmetric.NewMetrics(), err
 		}
-		buf := make([]byte, 4096)
-		n, err := c.Read(buf)
+		buf, err := io.ReadAll(c)
 		if err != nil {
 			return pmetric.NewMetrics(), err
 		}
-		buf = buf[0:n]
 		records, err = s.readStats(buf)
 		if err != nil {
 			return pmetric.NewMetrics(), fmt.Errorf("error reading stats: %w", err)

--- a/receiver/haproxyreceiver/scraper_test.go
+++ b/receiver/haproxyreceiver/scraper_test.go
@@ -41,6 +41,7 @@ func Test_scraper_readStats(t *testing.T) {
 			require.NoError(t, err2)
 			_, err2 = c.Write(stats)
 			require.NoError(t, err2)
+			require.NoError(t, c.Close())
 		default:
 			require.Fail(t, fmt.Sprintf("invalid message: %v", data))
 		}

